### PR TITLE
Restrict unique organization name for immediate child organizations under the parent organization

### PR DESCRIPTION
**Pull request by sadilchamishka**

### CODE REVIEW

File: components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationManagerImpl.java [hunk 1]
```diff
@@ -476,6 +479,20 @@ private void validateAddOrganizationRequest(Organization organization) throws Or
         validateAddOrganizationType(organization);
     }
 
+    private void validateSameOrgNameExistOnImmediateSubOrganizationsOfParentOrg(Organization organization)
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationManagerImpl.java:482]:
```suggestion
    private void validateOrgNameUniqunessAmongSiblings(Organization organization)
```

File: components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationManagerImpl.java [hunk 2]
```diff
@@ -476,6 +479,20 @@ private void validateAddOrganizationRequest(Organization organization) throws Or
...
+            throws OrganizationManagementException {
+
+        boolean foundSimilarOrganization =
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationManagerImpl.java:485]:
```suggestion
        boolean hasSigblingWithSameName =
```

File: components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationManagerImpl.java [hunk 3]
```diff
@@ -476,6 +479,20 @@ private void validateAddOrganizationRequest(Organization organization) throws Or
...
+                organizationManagementDAO.getChildOrganizations(organization.getParent().getId().trim(), false).stream()
+                        .anyMatch(basicOrganization ->
+                                StringUtils.equals(basicOrganization.getName(), organization.getName().trim()));
+        if (foundSimilarOrganization) {
+            throw handleClientException(ERROR_CODE_SAME_ORG_NAME_ON_IMMEDIATE_SUB_ORGANIZATIONS_OF_PARENT_ORG,
+                    organization.getName(), organization.getDescription());
+        }
+
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationManagerImpl.java:493]: remove additional new line

File: components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationManagerImpl.java (hunk 1 above)
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationManagerImpl.java:482]: What if we change this method signature to pass, organizationName, and the parent organization id.?

File: components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/constant/OrganizationManagementConstants.java [hunk 4]
```diff
@@ -272,6 +272,9 @@ public enum ErrorMessages {
         ERROR_CODE_INVALID_DELETE_SHARE_REQUEST("60069", "Invalid request.", "Cannot unshare " +
                 "the application with ID %s from the organization with ID %s if the application is shared with all " +
                 "children organizations."),
+        ERROR_CODE_SAME_ORG_NAME_ON_IMMEDIATE_SUB_ORGANIZATIONS_OF_PARENT_ORG("60070",
+                "Unable to create the organization due to same organization name found on immediate sub organizations of parent organization",
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/constant/OrganizationManagementConstants.java:276]:
```suggestion
                "Gien organization name is taken from a sibling organization.",
```

File: components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/constant/OrganizationManagementConstants.java [hunk 5]
```diff
@@ -272,6 +272,9 @@ public enum ErrorMessages {
...
+                "Organization with same organization name: %s found in the immediate sub organizations of parent organization"),
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/constant/OrganizationManagementConstants.java:277]:
```suggestion
                "Organization with same organization name: %s found in the immediate sub organizations of parent organization."),
```

File: components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationManagerImpl.java (hunk 1 above)
Comment (sadilchamishka) [reply components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationManagerImpl.java]: Fixed by - https://github.com/wso2-extensions/identity-organization-management-core/pull/29/commits/2d2707a5a9061170fb00f35525013252425a8586

File: components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationManagerImpl.java (hunk 2 above)
Comment (sadilchamishka) [reply components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationManagerImpl.java]: Fixed by - https://github.com/wso2-extensions/identity-organization-management-core/pull/29/commits/2d2707a5a9061170fb00f35525013252425a8586

File: components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationManagerImpl.java (hunk 3 above)
Comment (sadilchamishka) [reply components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationManagerImpl.java]: Fixed by - https://github.com/wso2-extensions/identity-organization-management-core/pull/29/commits/2d2707a5a9061170fb00f35525013252425a8586

File: components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationManagerImpl.java (hunk 1 above)
Comment (sadilchamishka) [reply components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationManagerImpl.java]: Fixed by - https://github.com/wso2-extensions/identity-organization-management-core/pull/29/commits/2d2707a5a9061170fb00f35525013252425a8586

File: components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/constant/OrganizationManagementConstants.java (hunk 4 above)
Comment (sadilchamishka) [reply components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/constant/OrganizationManagementConstants.java]: Fixed by -  https://github.com/wso2-extensions/identity-organization-management-core/pull/29/commits/2d2707a5a9061170fb00f35525013252425a8586

File: components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/constant/OrganizationManagementConstants.java (hunk 5 above)
Comment (sadilchamishka) [reply components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/constant/OrganizationManagementConstants.java]: Fixed by - https://github.com/wso2-extensions/identity-organization-management-core/pull/29/commits/2d2707a5a9061170fb00f35525013252425a8586

File: components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationManagerImpl.java [hunk 6]
```diff
@@ -717,6 +733,7 @@ private void validateOrganizationPatchOperations(List<PatchOperation> patchOpera
                     throw handleClientException(ERROR_CODE_SUPER_ORG_RENAME, organizationId);
                 }
                 validateOrganizationNameField(value);
+                newOrgName = value;
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationManagerImpl.java:736]: No need to maintain another variable right ?
```suggestion
            Organization organization = organizationManagementDAO.getOrganization(organizationId);
            validateOrgNameUniquenessAmongSiblings(organization.getParent().getId(), newOrgName);
```

Comment (sadilchamishka) [reply components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationManagerImpl.java]: Fixed by - https://github.com/wso2-extensions/identity-organization-management-core/pull/29/commits/9153863550364ed9286154d0fda7bf0d75ac820b

**Review by AnuradhaSK — APPROVED** [review]

**Pull request merged**